src/metrics.rs

```diff
@@ -0,0 +1,177 @@
+
+/// Reads an input stream that contains dura logs and enriches them with more analytics-ready info
+/// like number of insertions & deletions. The result is written back out to an output stream.
+pub fn get_snapshot_metrics(
+    input: &mut dyn io::Read,
+    output: &mut dyn io::Write,
+) -> FlexResult<()> {
+    let mut reader = io::BufReader::new(input);
+    let mut writer = io::BufWriter::new(output);
+    let mut line: u64 = 0; // for printing better error messages
+    let mut repo_cache: HashMap<String, Rc<Repository>> = HashMap::new();
+    loop {
+        line += 1;
+        let mut input_line = String::new();
+        if reader.read_line(&mut input_line)? == 0 {
+            return Ok(());
+        }
+        match scrape_log(input_line) {
+            Ok(Some(mut output)) => {
+                scrape_git(&mut output, &mut repo_cache)?;
+                writeln!(&mut writer, "{}", output)?;
+            }
+            Ok(None) => {}
+            // Seems like a good way to report errors, idk...
+            Err(e) => eprintln!("line {}: {}", line, e),
+        }
+    }
+}
+
+/// Scrape information out of the snapshot log.
+fn scrape_log(line: String) -> serde_json::Result<Option<Value>> {
+    let input_val: Value = serde_json::from_str(line.as_str())?;
+    let mut output_val = Value::Object(Map::new());
+
+    if let Some(t) = input_val.get("time") {
+        output_val["time"] = t.clone();
+    }
+
+    if let Some(op_value) = input_val.get("fields").and_then(|f| f.get("operation")) {
+        match from_value(op_value.clone())? {
+            Operation::Snapshot {
+                repo,
+                op: Some(op),
+                error: _,
+                latency,
+            } => {
+                output_val["repo"] = Value::String(repo);
+                if let Some(latency) = Number::from_f64(latency as f64) {
+                    output_val["latency"] = Value::Number(latency);
+                }
+                output_val["dura_branch"] = Value::String(op.dura_branch);
+                output_val["commit_hash"] = Value::String(op.commit_hash);
+                output_val["base_hash"] = Value::String(op.base_hash);
+            }
+            _ => return Ok(None),
+        }
+    } else {
+        return Ok(None);
+    }
+
+    Ok(Some(output_val))
+}
+
+/// Use the info captured from scrape_log to open a repo and capture information about the commit
+///
+/// The repo_cache is retained between calls. This cache seems to cut runtime by 50% in a
+/// completely non-scientific measure. It still seems to take unexpectedly long, probably because
+/// it still has to open lots of files (for each commit & tree object) behind the scenes, and this
+/// is inherently not cache-able.
+fn scrape_git(
+    value: &mut Value,
+    repo_cache: &mut HashMap<String, Rc<Repository>>,
+) -> Result<(), git2::Error> {
+    if let Some(repo_path_value) = value.get("repo") {
+        let repo_path = repo_path_value.as_str().unwrap_or("");
```

Comment:
Does the `unwrap_or` mean it falls back to the working dir? This would benefit from a comment

Reply:
Comment deffo clears it up, but unsure if that is good practice. Is there any benefit to knowingly passing an error to git rather than immediately throwing an error in dura?

Reply:
yeah, I ended up removing the comment and throwing an explicit error